### DISCUSSION TITLE
Add the chromeos ebuild directly as a CI target (and cross compile target)

### DISCUSSION THREAD

**Discussion by bleungatchromium** [fwupd/fwupd#2820] [Ideas]
@djcampello 
It would be ideal if we could run the chromeos ebuild directly as a CI target, maybe even a cross compile CI target.  Any chance you guys can try to put that together?

_Originally posted by @superm1 in https://github.com/fwupd/fwupd/issues/2131#issuecomment-634230392_

**Comment by hughsie** [comment]
@djcampello is this something still interesting to you? It would certainly help us with making syre we don't break anything.

**Comment by djcampello** [comment]
Yes, this is something on my to-do list

**Comment by djcampello** [comment]
I created a docker image with the chromiumos environment starting with a ubuntu docker image and basically following the steps at https://chromium.googlesource.com/chromiumos/docs/+/master/developer_guide.md

With the ubuntu-chromiumos image I can do the following:

$ sudo docker run --privileged -ti ubuntu-chromiumos /bin/bash -c "cd /root/chromiumos; sudo -u docker /root/depot_tools/cros_sdk emerge-amd64-generic fwupd"

Few things from the command above:
(1) needs to run with --privileged for the cros_sdk environment to work
(2) cros_sdk cannot be run as root, so I had to create a docker user on the system
(3) /root/depot_tools and /root/chromiumos contain the depot_tools and chromiumos repo checkouts (see size below)
(4) I only setup one board "amd64-generic" to build fwupd for

Finally the image is huge!!
campello@derch:~$ sudo docker image ls
REPOSITORY          TAG       IMAGE ID       CREATED              SIZE
ubuntu-chromiumos   latest    1cf20bcfa809   About a minute ago   51.6GB

I'm not sure if the size of this image is something that would work out with the ci here.

**Reply by superm1** [reply]
Woah... That would be too big I think. Would if help to do shallow checkouts?

**Reply by djcampello** [reply]
I don't think so:

728M	/root/depot_tools
13G	/root/chromiumos/.repo/
17G	/root/chromiumos/chroot
20G	/root/chromiumos/src

and inside the chroot some notable directories:
5.7G	/root/chromiumos/chroot/opt
8.7G	/root/chromiumos/chroot/usr
1.5G	/root/chromiumos/chroot/var

plus the actual board chroot with only fwupd and its dependencies built:
721M	/root/chromiumos/chroot/build/amd64-generic/

I expect the shallow checkouts to help only to reduce the 13G of the .repo/ directory if any.

**Reply by hughsie** [reply]
Whoa, I think that's an order of magnitude too large to run for every pull request. Would there be any way of running some kind of "shadow CI" i.e. you have a cron job running on your box once per week or something?

**Reply by djcampello** [reply]
I can work to setup something like that in my workstation.

I am also hoping to start working on some automated testing on the Chromium OS side for our fork. It may be possible that as part of that effort I might be able to setup a task to pull the master branch from upstream too. I'll keep you posted here on any development on that side.

**Reply by superm1** [reply]
> Whoa, I think that's an order of magnitude too large to run for every pull request

Here's my remaining ideas.  Feel free to shoot them down if they're bad.

1. Maybe removing stuff from the SDK that really isn't needed to build fwupd.  For example I doubt any of `/opt` will be needed.
2. Just pull  the Chrome ebuild to use on a Gentoo docker container.  Hopefully that hits a majority of the cases.

>I am also hoping to start working on some automated testing on the Chromium OS side for our fork.

The main goal here is an earlier notification that a change in master will break something anywhere.  That's part of why we pulled in packaging for all the distros we can.  We fix the packaging as we go rather than catching something late.
So if you have something checking master "downstream" that sounds good too.

**Reply by djcampello** [reply]
2. I think this would defeat the purpose. We have many custom packages and at random versions compared to Gentoo. We also have many build scripts that strip portions of the packages making them behave differently (the whole MIME type database issue comes to mind)
